src/consumer/fetcher.js

```diff
@@ -0,0 +1,58 @@
+const EventEmitter = require('events')
+const flatten = require('../utils/flatten')
+
+/**
+ * Fetches data from all assigned nodes, waits for workerQueue to drain and repeats.
+ *
+ * @param {object} options
+ * @param {number[]} options.nodeIds
+ * @param {import('./workerQueue').WorkerQueue} options.workerQueue
+ * @param {(nodeId: number) => Promise<T[]>} options.fetch
+ * @template T
+ */
+const createFetcher = ({ nodeIds, workerQueue, fetch }) => {
+  const emitter = new EventEmitter()
+  let isRunning = false
+
+  const getNodeIds = () => nodeIds
+
+  const getWorkerQueue = () => workerQueue
+
+  const fetchNodes = async () => {
+    const batches = await Promise.all(nodeIds.map(nodeId => fetch(nodeId)))
+    return flatten(batches)
+  }
+
+  const start = async () => {
+    isRunning = true
+
+    try {
+      const batches = await fetchNodes()
+
+      if (isRunning) {
+        await workerQueue.push(...batches)
+      }
+    } catch (error) {
+      isRunning = false
+      emitter.emit('end')
+      throw error
+    }
+
+    if (isRunning) {
+      return start()
```

Comment:
Embarrassing, TIL node.js doesn't support TCO (anymore).

I personally don't like the onCrash solution for Runner, because the callback can be called from multiple places and it's not clear how the errors are merged. But I guess it's the only way to go in that case.